pkg/controller/volumeattachment.go

```diff
@@ -0,0 +1,232 @@
+	kubeClientSet "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/apis/azuredisk/v1alpha1"
+	azVolumeClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/azureutils"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type ReconcileVolumeAttachment struct {
+	client         client.Client
+	azVolumeClient azVolumeClientSet.Interface
+	kubeClient     kubeClientSet.Interface
+	// retryMap allows volumeAttachment controller to retry Get operation for AzVolumeAttachment in case the CRI has not been created yet
+	retryMap   map[string]int
+	retryMutex sync.RWMutex
+	namespace  string
+}
+
+var _ reconcile.Reconciler = &ReconcileVolumeAttachment{}
+
+func (r *ReconcileVolumeAttachment) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	var volumeAttachment storagev1.VolumeAttachment
+	if err := r.client.Get(ctx, request.NamespacedName, &volumeAttachment); err != nil {
+		if !errors.IsNotFound(err) {
+			klog.Errorf("failed to get VolumeAttachment (%s): %v", request.Name, err)
+			return reconcile.Result{Requeue: true}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	volumeAttachmentExists := true
+	if now := metav1.Now(); volumeAttachment.DeletionTimestamp.Before(&now) {
+		volumeAttachmentExists = false
+	}
+
+	volumeName := volumeAttachment.Spec.Source.PersistentVolumeName
+	if volumeName == nil {
+		return reconcile.Result{Requeue: false}, status.Error(codes.Aborted, fmt.Sprintf("PV name is set nil for VolumeAttachment (%s)", volumeAttachment.Name))
+	}
+	nodeName := volumeAttachment.Spec.NodeName
+	if err := r.AnnotateAzVolumeAttachment(ctx, volumeAttachment, azureutils.GetAzVolumeAttachmentName(*volumeName, nodeName), volumeAttachment.Name, volumeAttachmentExists); err != nil {
+		return reconcile.Result{Requeue: true}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileVolumeAttachment) AnnotateAzVolumeAttachment(ctx context.Context, volumeAttachment storagev1.VolumeAttachment, azVolumeAttachmentName, volumeAttachmentName string, volumeAttachmentExists bool) error {
+	var azVolumeAttachment v1alpha1.AzVolumeAttachment
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: r.namespace, Name: azVolumeAttachmentName}, &azVolumeAttachment); err != nil {
+		if !errors.IsNotFound(err) {
+			klog.Errorf("failed to get AzVolumeAttachment (%s): %v", azVolumeAttachmentName, err)
+			return err
+			// if the AzVolumeAttachment object has already been deleted, ignore
+		}
+
+		r.retryMutex.RLock()
+		numRetry, ok := r.retryMap[azVolumeAttachmentName]
+		r.retryMutex.RUnlock()
+		if !ok {
+			r.retryMutex.Lock()
+			r.retryMap[azVolumeAttachmentName] = 1
```

Comment:
There is a race condition here when another thread operates on the same volume attachment concurrently. After release the lock on line 94, the other thread will race to acquire the exclusive lock on line 96. The loser will overwrite the update made by the winner. You should use the [double-checked locking pattern](https://en.wikipedia.org/wiki/Double-checked_locking) here, at least. (Look at the C# example for clarity.)

Rather the acquiring the exclusive lock to update the map entry, use [`atomic.AddInt32`](https://pkg.go.dev/sync/atomic#AddInt32) to update the value.

Reply:
I have added the double-checked locking pattern. I left the lock for map entry updates as we want to avoid concurrent writes to the map struct.